sound/soc/intel/boards/sof_da7219.c

```diff
@@ -354,30 +495,57 @@ static int audio_probe(struct platform_device *pdev)
 	if (pdev->id_entry && pdev->id_entry->driver_data)
 		board_quirk = (unsigned long)pdev->id_entry->driver_data;
 
-	/* backward-compatible with existing devices */
-	if (board_quirk & SOF_MAX98360A_SPEAKER_AMP_PRESENT)
-		card_da7219.name = devm_kstrdup(&pdev->dev, "da7219max98360a",
-						GFP_KERNEL);
-	else if (board_quirk & SOF_MAX98373_SPEAKER_AMP_PRESENT)
-		card_da7219.name = devm_kstrdup(&pdev->dev, "da7219max",
-						GFP_KERNEL);
+	if (board_quirk & SOF_DA7219_JSL_BOARD) {
+		/* backward-compatible with existing devices */
+		if (board_quirk & SOF_MAX98360A_SPEAKER_AMP_PRESENT)
+			card_da7219.name = devm_kstrdup(&pdev->dev,
+							"da7219max98360a",
+							GFP_KERNEL);
+		else if (board_quirk & SOF_MAX98373_SPEAKER_AMP_PRESENT)
+			card_da7219.name = devm_kstrdup(&pdev->dev, "da7219max",
+							GFP_KERNEL);
+
+		dai_links = jsl_dais;
+		amp_idx = 0;
+
+		card_da7219.num_links = ARRAY_SIZE(jsl_dais);
+	} else {
+		dai_links = adl_dais;
+		amp_idx = 7;
+
+		card_da7219.num_links = ARRAY_SIZE(adl_dais);
+	}
 
 	dev_dbg(&pdev->dev, "board_quirk = %lx\n", board_quirk);
 
 	/* speaker amp */
 	if (board_quirk & SOF_MAX98360A_SPEAKER_AMP_PRESENT) {
-		max_98360a_dai_link(&dais[0]);
+		max_98360a_dai_link(&dai_links[amp_idx]);
 	} else if (board_quirk & SOF_MAX98373_SPEAKER_AMP_PRESENT) {
-		dais[0].codecs = max_98373_components;
-		dais[0].num_codecs = ARRAY_SIZE(max_98373_components);
-		dais[0].init = max_98373_spk_codec_init;
-		dais[0].ops = &max98373_ops; /* use local ops */
+		dai_links[amp_idx].codecs = max_98373_components;
+		dai_links[amp_idx].num_codecs = ARRAY_SIZE(max_98373_components);
+		dai_links[amp_idx].init = max_98373_spk_codec_init;
+		if (board_quirk & SOF_DA7219_JSL_BOARD) {
+			dai_links[amp_idx].ops = &max98373_ops; /* use local ops */
+		} else {
+			/* TBD: implement the amp for later platform */
+			dev_err(&pdev->dev, "max98373 not support yet\n");
+			return -EINVAL;
+		}
 
 		max_98373_set_codec_conf(&card_da7219);
 	}
 
+	card_da7219.dai_link = dai_links;
+
 	INIT_LIST_HEAD(&ctx->hdmi_pcm_list);
 
+	/* Use PLL bypass mode if MCLK is available, be sure to set the
+	 * frequency of MCLK to 12.288 or 24.576MHz on topology side.
+	 */
+	if (board_quirk & SOF_DA7219_PLL_BYPASS)
+		ctx->pll_bypass = 1;
```

Comment:
why is this needed? What not just test the MCLK and automatically bypass the PLL if it's 12.228 or a multiple?

Reply:
oh I can do it in codec init. let me upload a new version.